### PR TITLE
make fragment compatible for rms

### PR DESCRIPTION
**Pull request by lily90502**
Make rms being able to read & draw fragments

### CODE REVIEW

**Comment by hwpang** [comment]
I rebased and cleaned up unnecessary changes as @JacksonBurns and Anna need this to wrap up the fragment work.

**Review by JacksonBurns — APPROVED** [review]
LGTM - my only thought is that you might want to delete the commented lines, rather than carry the commented out version forward.

**Comment by hwpang** [comment]
@JacksonBurns Thanks for catching that, I didn't notice that Lily commented those lines out. Those are changes for her personal use and should not be included in this PR. I edited that commit.

**Comment by JacksonBurns** [comment]
@hwpang sure thing! I will re-run the tests since (from visual inspection) the errors were possibly spurious. If they fail again for the same reason, could you take a look? Tests which seem unrelated to this PR are failing.

**Comment by hwpang** [comment]
@JacksonBurns Thanks for rerunning the tests! The CI still failed after rerunning, due to the following error:

```
┌ Warning: Error requiring `EnzymeCore` from `KernelAbstractions`
│   exception =
│    LoadError: UndefVarError: EnzymeRules not defined`
```

I suspect that the `EnzymeCore v0.1.0` and `Enzyme v0.10.18` used during installation is not compatible with `KernelAbstractions v0.9.8`. I see that there is a `Enzyme v0.11.8`, so I'm attempting to relax package specification to see if that can be solved.

**Comment by hwpang** [comment]
I realize that the reason the CI test failed had nothing to do with `EnzymeCore` or `Enzyme` packages, but that the adjoint sensitivity results are different from the forward sensitivity results as mentioned in https://github.com/ReactionMechanismGenerator/ReactionMechanismSimulator.jl/issues/235. @JacksonBurns could you re-review the PR? I think the failed CI test is not related to this PR.

**Review by JacksonBurns — DISMISSED** [review]
LGTM! Thanks for figuring out the issue with the CI

File: src/Simulation.jl [hunk 1]
```diff
@@ -319,7 +319,7 @@ end
 function rops(ssys::SystemSimulation, name, t)
     domains = getfield.(ssys.sims, :domain)
     ind = findfirst(isequal(name), ssys.names)
-    Nrxns = sum([length(sim.domain.phase.reactions) for sim in ssys.sims]) + sum([length(inter.reactions) for inter in ssys.interfaces if hasproperty(inter, :reactions)])
+    Nrxns = sum([length(sim.domain.phase.reactions) for sim in ssys.sims]) + sum(Vector{Int}([length(inter.reactions) for inter in ssys.interfaces if hasproperty(inter, :reactions)]))
```
Comment (mjohnson541) [inline src/Simulation.jl:322]: Can the commit message here be more specific?

File: Project.toml [hunk 2]
```diff
@@ -40,32 +40,32 @@ YAML = "ddb6d928-2868-570f-bddf-ab3f9cf99eb6"
 Zygote = "e88e6eb3-aa80-5325-afca-941959d7151f"
 
 [compat]
-Calculus = "0.4,0.5"
-Colors = "0.11,0.12"
-Conda = "1"
+Calculus = "^0.5"
+Colors = "^0.12"
+Conda = "^1"
```
Comment (mjohnson541) [inline Project.toml:45]: The "^" is default behavior, so adding the caret when there isn't another operator there doesn't change behavior.

**Review by mjohnson541 — COMMENTED** [review]
Thanks! 

There are a few try-catches here that aren't for specific errors...that could be annoying for users...suppose someone messes up a Molecule adjlist...they won't get the actual error associated with constructing a Molecule, they will get the error associated with constructing a Fragment. 

Also is it possible to add the bugfixes for https://github.com/ReactionMechanismGenerator/ReactionMechanismSimulator.jl/issues/219 to this PR?

Other than that just a couple comments.

**Comment by JacksonBurns** [comment]
@hwpang can you take care of the comments from @mjohnson541's review? I'm not sure what the type cast (if it is a type cast) exactly accomplishes

**Comment by hwpang** [comment]
@JacksonBurns Yes, I will work on this! Unfortunately I have commitments today and tomorrow, so the earliest I can work on this is Friday. In the meanwhile, you and Anna can use this branch for now for your fragment related work.

**Comment by hwpang** [comment]
@mjohnson541 I addressed your comments in the new push! I have an attempt to fix the issue (https://github.com/ReactionMechanismGenerator/ReactionMechanismSimulator.jl/issues/219) in a previously opened PR (https://github.com/ReactionMechanismGenerator/ReactionMechanismSimulator.jl/pull/230). Can you also review that PR? I think It currently failed one of the new unit tests I added. I will try getting around to fix it soon. I prefer to have separate PRs and each solves a small bug, rather than a larger PR that solves all the bug.

File: Project.toml (hunk 2 above)
Comment (hwpang) [reply Project.toml]: I dropped this commit!

File: src/fluxdiagrams.jl [hunk 3]
```diff
@@ -42,11 +42,19 @@ function drawspc(spc::Species,path::String=".")
         end
     end
     if spc.adjlist != ""
-        mol = molecule.Molecule().from_adjacency_list(spc.adjlist)
```
Comment (mjohnson541) [inline src/fluxdiagrams.jl:45]: Can we do this one with Species the same way you do later?

**Review by mjohnson541 — COMMENTED** [review]
Looks mostly good, one last thing.

File: src/fluxdiagrams.jl (hunk 3 above)
Comment (hwpang) [reply src/fluxdiagrams.jl:45]: Yep! Done.

**Comment by hwpang** [comment]
It turns out you can't check if a Python object is of a Python class instance in Julia, which causes some part of tests to fail because `isa` cannot take a Python object as type. I am not sure what is the best way forward. @mjohnson541 Would love your thoughts on this.

**Comment by hwpang** [comment]
It turns out that the problem was with `rmgmolecule`. The conformer attribute was commented out in this [commit](https://github.com/ReactionMechanismGenerator/molecule/commit/120dff95518deb50b6a3a69887c9f8dd9fe21f3a#diff-7d985f3f81703ccea15b36cda23fe11b37e0825ecc920ea7502e79777e4bd5aeR45), but its counterpart in the `.py` file was not. This causes the initialization of `spc = Species()` to fail with `AttributeError("'molecule.species.Species' object has no attribute 'conformer'")` error. I work around to not use `Species` in the new push.

**Comment by hwpang** [comment]
The newest changes have passed the tests that we are able to pass currently. The two tests failed are related to https://github.com/ReactionMechanismGenerator/ReactionMechanismSimulator.jl/issues/235. I will clean up the commits a bit.

**Comment by hwpang** [comment]
@mjohnson541 All comments are addressed, please re-review!

**Review by mjohnson541 — APPROVED** [review]

**Pull request merged**